### DISCUSSION TITLE
Eglot always trigger a full rebuild in Rust project?

### DISCUSSION THREAD

**Discussion by azzamsa** [joaotavora/eglot#1250] [Q&A] [answered]
Hi

Using rust-mode/rustic-mode with Eglot always triggers a full rebuild in the Rust project.
This is not the case with VSCode as shown in the video below.

However, this is not the case if I start Emacs from the terminal.

Any suggestions on how to debug this? `eglot-events-buffer` doesn't help me.

Thanks!

https://github.com/abougouffa/minemacs/assets/17734314/86efcc58-5e5c-4409-9421-3e3ff16ec318

**Comment by joaotavora** [comment]
Please try to reproduce this with only Emacs + Eglot.  In my toy rust project experimentation with Emacs + Eglot (+ rust-mode, not rustic-mode) I don't think I see this.  

See https://joaotavora.github.io/eglot/#Troubleshooting-Eglot for techniques on how to submit MRE reproduction recipes.  It takes some work, but is usually helpful in pinpointing the problem.

Else you may hope that someone with a similar setup spots this and helps you out.

**Reply by joaotavora** [reply]
After having watched the video, I'm not even sure what I'm looking at.  What is the pane on the left?  is this a watcher process that does things when it notices files are changed/saved.  Be sure to describe this in the MRE recipe.

**Reply by azzamsa** [reply]
On the left side is just a `cargo watch -x 'clippy --all-targets --all-features'` command. I tried to run clippy manually `clippy --all-targets --all-features'`. The behavior is the same. With Eglot it needs a full rebuild, but not with VSCode. Even if both of them use the same `rust-analyzer`.

As I said, the strange thing is that the behavior would be similar between VSCode and Eglot if I run Emacs from the terminal.

It is my first week using my custom configuration, previously I use Doom, and I don't find this behavior in Doom. I will investigate further.

**Reply by joaotavora** [reply]
I'm not very familiar with Rust, so you need to describe -- textually, and without assuming anything else other than a basic Rust toolchain and a recent version of Emacs vanilla (which already includes Eglot) -- how to witness your behaviour first hand.  See https://joaotavora.github.io/eglot/#Troubleshooting-Eglot and follow the 7 steps outlined. I have `rust`, `cargo` and `rust-analyzer` installed on a linux machine.  I don't know what clippy is but you can explain as part of those 7 steps.  

Else I don't know what to look for and I don't have time to go on a wild good chase and guess what the problem might be.

**Reply by azzamsa** [reply, marked as answer]
I found the root cause https://github.com/azzamsa/dotfiles/commit/f2df26e6f211f4782938edc0a39e56419017c9e5

> Emacs eglot-mode and lsp-mode always rebuild the entire Rust/Cargo project upon
changes. I assume that Emacs uses different Rust tooling (`rust-analyzer`, `cargo`, etc)
compared to the one that I use from the terminal. Because `rustup` install both tools in `~/.cargo`
and `rustup/<target>` directory. `This force `cargo` to rebuild the entire project.

> After resetting the `cargo` and `rustup` path to default. The issue is solved.

> This doesn't apply to VSCodium, because it uses its own rust-analyzer.

I move all of them because of https://github.com/b3nj5m1n/xdg-ninja, I think I need to make them aware of this.

**Reply by joaotavora** [reply]
Nice, thanks for analyzing.  And yes -- if you don't specifically tell it otherwise -- Emacs/Eglot will generally use the system tooling that is available in `PATH`
